challenges/week4.cinema-seat-allocation/puchi.js

```diff
@@ -0,0 +1,55 @@
+// https://leetcode.com/problems/cinema-seat-allocation/
+
+// Runtime: 144 ms, faster than 68.75% of JavaScript online submissions for Cinema Seat Allocation.
+// Memory Usage: 46.1 MB, less than 87.50% of JavaScript online submissions for Cinema Seat Allocation.
+
+const maxNumberOfFamilies = (n, reserved) => {
+  // 4인석의 갯수
+  let families = 0;
+  // 검사한 줄, 지난 행
+  let [checkedRows, lastRow] = [1, 1];
+  // 2345, 4567, 6789가 각각 비어 있는지 여부
+  let [isLeftEmpty, isMiddleEmpty, isRightEmpty] = [true, true, true];
+
+  // 예약석을 행(오름차순) + 열(오름차순) 순으로 정렬하고,
+  // 예약석이 있는 줄에서 4인석의 갯수를 계산함
+  reserved
+    .sort(([aRow, aCol], [zRow, zCol]) => (aRow === zRow ? aCol - zCol : aRow - zRow))
+    .forEach(([row, col]) => {
+      if (row !== lastRow) {
+        families += (isLeftEmpty || isMiddleEmpty || isRightEmpty) + (isLeftEmpty && isRightEmpty);
+        checkedRows += 1;
+        lastRow = row;
+        [isLeftEmpty, isMiddleEmpty, isRightEmpty] = [true, true, true];
+      }
+      if (isLeftEmpty) isLeftEmpty = offRange(col, 2, 5);
+      if (isMiddleEmpty) isMiddleEmpty = offRange(col, 4, 7);
+      if (isRightEmpty) isRightEmpty = offRange(col, 6, 9);
+    });
+
```

Comment:
처음에는 어떻게 동작하는지 잘 몰랐는데, `checkedRows` 아이디어가 좋네요. 다만 `forEach`를 사용해서 루프를 함수 call 형태로 사용했지만, 각 루프마다 적용되는 state가 2개나 있다는 점이 읽기 어색하게 만드는 것 같습니다. 아래처럼 고전적인 for 구문을 사용하도록 고친다면 어땠을까요?
(속도 면에서 어느 쪽이 더 장점이 있는지는 모르겠습니다)

```suggestion
  reserved
    .sort(([aRow, aCol], [zRow, zCol]) => (aRow === zRow ? aCol - zCol : aRow - zRow))
  
  for (const [row, col] in reserved) {
    if (row !== lastRow) {
      families += (isLeftEmpty || isMiddleEmpty || isRightEmpty)
                + (isLeftEmpty && isRightEmpty);
      checkedRows += 1;
      lastRow = row;
      [isLeftEmpty, isMiddleEmpty, isRightEmpty] = [true, true, true];
    }
    if (isLeftEmpty) isLeftEmpty = offRange(col, 2, 5);
    if (isMiddleEmpty) isMiddleEmpty = offRange(col, 4, 7);
    if (isRightEmpty) isRightEmpty = offRange(col, 6, 9);
  }
```

Reply:
오 그렇군여 루프 틱마다 클로저에 의해 실행 스택에 들어가는 쓸데없는 데이터가 있으니 반복문이 좋을 거 같네요. 역시 짬뽕 언어는 이런 부분에서 짬뽕스럽게 짜주지 않으면! 🤣

`for in`은 객체 리터럴의 키 모음에 대한 루프니까 `for of`로 고쳐서 제출해 봤습니다. (`for of`가 forEach`보다는 약간 느리지만 신경쓸 레벨은 아니라고 생각하므로, 속도 이외에 합당한 이유가 있는 쪽을 쓰고 싶네요)

Reply:
앗 `for of`... 검색해 보다가 나온 키워드였는데 귀찮아서 안눌러본게 이렇게 연결되는군욬ㅋㅋㅋㅋ

Reply:
브레인스토밍 과정은 즐거움이죠 껄껄 이 즐거움을 위한 프로젝트가 아니겠습니까 (대충 턱괴는짤)